File: examples/multi-lang/ruby/mcp_client.rb
#!/usr/bin/env ruby
# MCP Python SDK - Ruby Integration Example
# This script demonstrates how to interact with MCP servers from Ruby

require 'json'
require 'open3'

class McpClient
  attr_reader :server_process
  
  def initialize(server_command, server_args = [])
    @server_command = server_command
    @server_args = server_args
    @server_process = nil
  end
  
  # Start the MCP server process
  def start_server
    puts "Starting MCP server..."
    @stdin, @stdout, @stderr, @wait_thr = Open3.popen3(@server_command, *@server_args)
    puts "✓ Server started with PID: #{@wait_thr.pid}"
  end
  
  # Send a request to the MCP server
  def send_request(method, params = {})
    request = {
      jsonrpc: "2.0",
      id: rand(1..1000),
      method: method,
      params: params
    }
    
    request_json = request.to_json
    puts "Sending request: #{request_json}"
    
    @stdin.puts(request_json)
    @stdin.flush
    
    # Read response
    response_line = @stdout.gets
    return nil if response_line.nil?
    
    response = JSON.parse(response_line)
    puts "Received response: #{response}"
    response
  end
  
  # Initialize the MCP connection
  def initialize_connection
    send_request("initialize", {
      protocolVersion: "2024-11-05",
      capabilities: {
        tools: {},
        resources: {},
        prompts: {}
      },
      clientInfo: {
        name: "ruby-mcp-client",
        version: "1.0.0"
      }
    })
  end
  
  # List available tools
  def list_tools
    response = send_request("tools/list")
    if response && response["result"]
      puts "\nAvailable tools:"
      response["result"]["tools"].each do |tool|
        puts "  - #{tool['name']}: #{tool['description']}"
      end
    end
    response
  end
  
  # Call a tool
  def call_tool(tool_name, arguments = {})
    send_request("tools/call", {
      name: tool_name,
      arguments: arguments
    })
  end
  
  # List available resources
  def list_resources
    response = send_request("resources/list")
    if response && response["result"]
      puts "\nAvailable resources:"
      response["result"]["resources"].each do |resource|
        puts "  - #{resource['uri']}: #{resource['name']}"
      end
    end
    response
  end
  
  # Read a resource
  def read_resource(uri)
    send_request("resources/read", { uri: uri })
  end
  
  # Stop the server
  def stop_server
    if @server_process
      puts "Stopping MCP server..."
      Process.kill('TERM', @wait_thr.pid)
      @stdin.close
      @stdout.close
      @stderr.close
      puts "✓ Server stopped"
    end
  end
end

# Example usage
class McpExample
  def self.run
    puts "==================================="
    puts "MCP Python SDK - Ruby Integration"
    puts "==================================="
    
    # Create client for an MCP server
    client = McpClient.new("uv", ["run", "python", "-m", "mcp"])
    
    begin
      # Start the server
      client.start_server
      sleep 1 # Give server time to start
      
      # Initialize connection
      puts "\nInitializing MCP connection..."
      client.initialize_connection
      
      # List available tools
      client.list_tools
      
      # Example: Call a tool (if available)
      # response = client.call_tool("example_tool", { input: "test" })
      # puts "Tool response: #{response}"
      
      # List resources
      client.list_resources
      
      puts "\n✓ Integration example completed successfully!"
      
    rescue => e
      puts "Error: #{e.message}"
      puts e.backtrace
    ensure
      # Clean up
      client.stop_server
    end
  end
  
  def self.create_example_server
    server_code = <<~PYTHON
      """
      Example MCP Server
      Created by Ruby script
      """
      from mcp.server.fastmcp import FastMCP
      
      mcp = FastMCP("Ruby Example Server")
      
      @mcp.tool()
      def echo(message: str) -> str:
          """Echo a message back"""
          return f"Echo: {message}"
      
      @mcp.tool()
      def process_data(data: dict) -> dict:
          """Process JSON data"""
          return {"processed": True, "input": data}
      
      if __name__ == "__main__":
          mcp.run()
    PYTHON
    
    File.write("ruby_example_server.py", server_code)
    puts "✓ Created example server: ruby_example_server.py"
  end
end

# Run the example if this file is executed directly
if __FILE__ == $0
  McpExample.run
end

File: examples/multi-lang/kotlin/McpClient.kt
package io.modelcontextprotocol.example

import java.io.*
import java.util.concurrent.TimeUnit

/**
 * MCP Python SDK - Kotlin Integration Example
 * This class demonstrates how to interact with MCP servers from Kotlin/JVM
 */
class McpClient(
    private val serverCommand: String,
    private val serverArgs: List<String> = emptyList()
) {
    private var process: Process? = null
    private var writer: BufferedWriter? = null
    private var reader: BufferedReader? = null
    
    /**
     * Start the MCP server process
     */
    fun startServer() {
        println("Starting MCP server...")
        val command = mutableListOf(serverCommand).apply { addAll(serverArgs) }
        
        val processBuilder = ProcessBuilder(command)
            .redirectErrorStream(true)
        
        process = processBuilder.start()
        
        writer = BufferedWriter(OutputStreamWriter(process!!.outputStream))
        reader = BufferedReader(InputStreamReader(process!!.inputStream))
        
        println("✓ Server started")
    }
    
    /**
     * Send a JSON-RPC request to the MCP server
     */
    fun sendRequest(method: String, params: Map<String, Any> = emptyMap()): String? {
        val request = buildJsonRpcRequest(method, params)
        println("Sending request: $request")
        
        writer?.write(request)
        writer?.newLine()
        writer?.flush()
        
        val response = reader?.readLine()
        println("Received response: $response")
        return response
    }
    
    /**
     * Build a JSON-RPC request
     */
    private fun buildJsonRpcRequest(method: String, params: Map<String, Any>): String {
        val paramsJson = params.entries.joinToString(",") { (k, v) -> 
            "\"$k\":${toJsonValue(v)}"
        }
        
        return """{
            "jsonrpc":"2.0",
            "id":${(1..1000).random()},
            "method":"$method",
            "params":{$paramsJson}
        }"""
    }
    
    private fun toJsonValue(value: Any): String = when (value) {
        is String -> "\"$value\""
        is Number -> value.toString()
        is Boolean -> value.toString()
        is Map<*, *> -> {
            val entries = value.entries.joinToString(",") { (k, v) ->
                "\"$k\":${toJsonValue(v!!)}"
            }
            "{$entries}"
        }
        else -> "\"$value\""
    }
    
    /**
     * Initialize the MCP connection
     */
    fun initializeConnection(): String? {
        return sendRequest("initialize", mapOf(
            "protocolVersion" to "2024-11-05",
            "capabilities" to mapOf(
                "tools" to emptyMap<String, Any>(),
                "resources" to emptyMap<String, Any>(),
                "prompts" to emptyMap<String, Any>()
            ),
            "clientInfo" to mapOf(
                "name" to "kotlin-mcp-client",
                "version" to "1.0.0"
            )
        ))
    }
    
    /**
     * List available tools
     */
    fun listTools(): String? {
        return sendRequest("tools/list")
    }
    
    /**
     * Call a tool
     */
    fun callTool(toolName: String, arguments: Map<String, Any> = emptyMap()): String? {
        return sendRequest("tools/call", mapOf(
            "name" to toolName,
            "arguments" to arguments
        ))
    }
    
    /**
     * List available resources
     */
    fun listResources(): String? {
        return sendRequest("resources/list")
    }
    
    /**
     * Read a resource
     */
    fun readResource(uri: String): String? {
        return sendRequest("resources/read", mapOf("uri" to uri))
    }
    
    /**
     * Stop the server
     */
    fun stopServer() {
        println("Stopping MCP server...")
        writer?.close()
        reader?.close()
        process?.destroy()
        process?.waitFor(5, TimeUnit.SECONDS)
        println("✓ Server stopped")
    }
}

/**
 * Example usage demonstrating MCP integration from Kotlin
 */
object McpExample {
    @JvmStatic
    fun main(args: Array<String>) {
        println("===================================")
        println("MCP Python SDK - Kotlin Integration")
        println("===================================")
        
        val client = McpClient("uv", listOf("run", "python", "-m", "mcp"))
        
        try {
            // Start the server
            client.startServer()
            Thread.sleep(1000) // Give server time to start
            
            // Initialize connection
            println("\nInitializing MCP connection...")
            client.initializeConnection()
            
            // List available tools
            println("\nListing tools...")
            client.listTools()
            
            // Example: Call a tool (if available)
            // val response = client.callTool("example_tool", mapOf("input" to "test"))
            // println("Tool response: $response")
            
            // List resources
            println("\nListing resources...")
            client.listResources()
            
            println("\n✓ Integration example completed successfully!")
            
        } catch (e: Exception) {
            println("Error: ${e.message}")
            e.printStackTrace()
        } finally {
            // Clean up
            client.stopServer()
        }
    }
    
    /**
     * Create an example MCP server file
     */
    fun createExampleServer() {
        val serverCode = """
            """
            Example MCP Server
            Created by Kotlin script
            """
            from mcp.server.fastmcp import FastMCP
            
            mcp = FastMCP("Kotlin Example Server")
            
            @mcp.tool()
            def process_text(text: str, operation: str = "uppercase") -> str:
                '''Process text with various operations'''
                if operation == "uppercase":
                    return text.upper()
                elif operation == "lowercase":
                    return text.lower()
                elif operation == "reverse":
                    return text[::-1]
                else:
                    return text
            
            @mcp.tool()
            def calculate_sum(numbers: list) -> float:
                '''Calculate sum of numbers'''
                return sum(numbers)
            
            @mcp.resource("config://app-settings")
            def get_settings() -> str:
                '''Get application settings'''
                return '{"language": "kotlin", "version": "1.9", "jvm": "17"}'
            
            if __name__ == "__main__":
                mcp.run()
        """.trimIndent()
        
        File("kotlin_example_server.py").writeText(serverCode)
        println("✓ Created example server: kotlin_example_server.py")
    }
}
